wagtail/admin/views/chooser.py

```diff
@@ -642,40 +643,45 @@ def post(self, request):
             # Otherwise this could get very annoying accidentally matching coincidentally
             # named pages on different sites
 
+            possible_sites = defaultdict(list)
+
             if match_relative_paths:
-                possible_sites = []
                 for pk, path, url, language_code in sites:
-                    possible_sites.append((pk, url_without_query))
+                    possible_sites[pk].append(url_without_query)
 
                     # If the submitted URL is prefixed with the serve path,
                     # also consider it without the serve path so we can match
                     # the page using Page.route()
                     if serve_path and url_without_query.startswith(serve_path):
-                        possible_sites.append(
-                            (pk, url_without_query[len(serve_path) - 1 :])
+                        possible_sites[pk].append(
+                            url_without_query[len(serve_path) - 1 :]
                         )
             else:
-                possible_sites = []
                 for pk, path, url, language_code in sites:
                     if not submitted_url.startswith(url):
                         continue
-                    possible_sites.append((pk, url_without_query[len(url) :]))
+                    possible_sites[pk].append(url_without_query[len(url) :])
 
                     # If the submitted URL is prefixed with the serve path,
                     # also consider it without the serve path so we can match
                     # the page using Page.route()
                     if serve_path and url_without_query.startswith(url + serve_path):
-                        possible_sites.append(
-                            (pk, url_without_query[len(url) + len(serve_path) - 1 :])
+                        possible_sites[pk].append(
+                            url_without_query[len(url) + len(serve_path) - 1 :]
                         )
 
             # Loop over possible sites to identify a page match
-            for pk, url in possible_sites:
-                try:
-                    route = Site.objects.get(pk=pk).root_page.specific.route(
-                        request,
-                        [component for component in url.split("/") if component],
-                    )
+            for pk, possible_urls in possible_sites.items():
+                site = Site.objects.select_related("root_page").get(pk=pk)
+                root_page = site.root_page.specific
+                for url in possible_urls:
+                    try:
+                        route = root_page.route(
+                            request,
+                            [component for component in url.split("/") if component],
+                        )
+                    except Http404:
+                        continue
```

Comment:
As far as I understand, `Http404` will only be raised by `route()` in this block, so I kept this simple by moving the `except Http404` all the way up here instead of encompassing the whole conversion logic. This also helps avoid unnecessary whitespace changes after I changed the structure to use a `dict[str, list]` which meant I had to do a nested loop here.